### PR TITLE
Add wrappers for TorchRL training workflow 

### LINKED ISSUE #1181: [Proposal] Add TorchRL to IsaacLab workflows
### Proposal

[TorchRL ](https://pytorch.org/rl/stable/index.html) is a RL framework produced by the PyTorch team that is growing in popularity within the research community. It offers fast optimization times, flexibility, and leverages tensordicts for data passing between environment and algorithm. It would be nice to add TorchRL as a new workflow option to IsaacLab. 

### Motivation

Please outline the motivation for the proposal. Summarize the core use cases and user problems and needs you are trying to solve.

Is your feature request related to a problem? e.g.,"I'm always frustrated when [...]".

If this is related to another GitHub issue, please link here too.

### Alternatives

A clear and concise description of any alternative solutions or features you've considered, if any.

### Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)

### Acceptance Criteria

Add the criteria for which this task is considered **done**. If not known at issue creation time, you can add this once the issue is assigned.

- [ ] Complete TorchRL train and play pipeline integrated into IsaacLab.

### PR DESCRIPTION
**Pull request by fyu-bdai**
# Description

Adds TorchRL module wrappers, PPO runner, and PPO runner cfg for training IsaacLab environments with TorchRL.  

This PR is the first in a series of three that together, adds a complete training pipeline for the Anymal-D training environment using torchrl. This PR contains core wrapper modules that should be merged first.  

Related PRs:
#1179 Adds torchrl play and train scripts. 
#1180 Adds Anymal-D torchrl training configuration. 

Fixes #1181 

## :warning: Not Supported ##

- Empirical normalization.
- Recurrent networks.
- Training with Neptune logger.

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

### CODE REVIEW

**Comment by fyu-bdai** [comment]
Can [Vincent Moens](https://github.com/vmoens) also get tagged as a reviewer?

**Comment by fyu-bdai** [comment]
> In my opinion, non-wrapper code (e.g.: `torchrl` runner code) should be in the `source/standalone/workflows` folder.
> 
> ```
> source/standalone/workflows
> ├── torchrl
> │   ├── ppo/      <-- code here
> │   ├── cli_args.py
> │   ├── play.py
> │   └── train.py
> ```
> 
> Having the PPO runner code in extension will generate non-Isaac Lab related version changes and changelog records when the code need to be modified.

I realized this causes issues with the environment specific PPO configurations, which needs to import the TorchRL PPO runner cfgs. I can move only the configs to `omni.isaac.lab_tasks.utils.wrappers`, but that would split the location of the torchrl runner and torchrl runner cfg. I have elected to move them back as they are for now.

**Comment by jsmith-bdai** [comment]
Hey Fangzhou, can you clarify why a single environment doesn't work?

**Comment by fyu-bdai** [comment]
> Hey @fyu-bdai , can you clarify why a single environment doesn't work?

I took another look again and fixed the single env training bug with TorchRL.

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_ppo_runner.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+import torch
+import yaml
+from numbers import Real
+from tensordict.nn import InteractionType, TensorDictModule
+from torch import nn
+from torch.distributions import Normal
+from typing import TYPE_CHECKING
+
+import wandb
+from torchrl.data.tensor_specs import CompositeSpec, UnboundedContinuousTensorSpec
```
Comment (vmoens) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_ppo_runner.py:24]: You may want to use `Composite` and `Unbounded` - we're deprecating `UnboundedContinuousTensorSpec`

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_ppo_runner.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+from torchrl.envs.libs.gym import GymEnv
```
Comment (vmoens) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_ppo_runner.py:25]:
```suggestion
from torchrl.envs import GymEnv
```

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+        Args:
+            tensordict (TensorDictBase): a tensordict coming from the output of :func:`~torchrl.envs.utils.step_mdp`.
+
+        Returns:
+            A tensordict that is identical to the input where the environment was
+            not reset and contains the new reset data where the environment was reset.
+
+        """
+        if self._simple_done:
+            done = tensordict._get_str("done", default=None)
```
Comment (vmoens) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py:57]:
```suggestion
            done = tensordict.get("done", default=None)
```
could be a bit more robust to bc-breaking changes (even though I don't expect any). The overhead is low and can be swallowed by compile

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/exporter.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+# Copyright (c) 2022-2025, The Isaac Lab Project Developers.
```
Comment (vmoens) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/exporter.py:1]: you probably want to look at these tutos
https://pytorch.org/rl/main/tutorials/export.html
https://pytorch.org/tensordict/main/tutorials/export.html

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,601 @@
...
+        super().__init__(env, categorical_action_encoding=categorical_action_encoding, **kwargs)
+        self._ep_len_buf = torch.zeros(env.unwrapped.num_envs, device=env.unwrapped.device)
+        self._ep_reward_buf = torch.zeros(env.unwrapped.num_envs, device=env.unwrapped.device)
+        self._curr_ep_len = torch.zeros(env.unwrapped.num_envs, device=env.unwrapped.device)
+        self._curr_reward_sum = torch.zeros(env.unwrapped.num_envs, device=env.unwrapped.device)
+        self._env_reset_mask = torch.zeros(env.unwrapped.num_envs, device=env.unwrapped.device, dtype=torch.bool)
+
+    def maybe_reset(self, tensordict: TensorDictBase) -> TensorDictBase:
+        """Checks the done keys of the input tensordict. Unlike the base GymWrapper implementation, we do not
+        call env.reset()
```
Comment (vmoens) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py:46]: interesting, we bumped into a similar problem with @luisenp and @teopir

We need a better handling of auto-resetting envs in torchrl.

How do you get the last values of a trajectory (if you can get them)?

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+            # the necessary keys
+            next_tensordict.update(next_preset.exclude(*next_tensordict.keys(True, True)))
+        tensordict.set("next", next_tensordict)
+        # Information from the "extras" dict of dicts passed from IsaacLab contains reward and metrics information.
+        # The _data postfix is added to distinguish the field containing the actual values from the dummy {key} field
+        # in the tensordict required for faster rollouts. To get logging to work, we need to move the data from the "next" tensordict
+        # to the top level tensordict.
+        # TODO: Find a more elegant solution to enable rewards and metrics logging. "
+        for key in self.env.unwrapped.extras:
+            tensordict[f"{key}_data"] = tensordict["next"][f"{key}_data"]
```
Comment (vmoens) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py:112]: I don't grasp why the reward needs to be brought to the root - can you explain?

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,601 @@
...
+        return tensordict
+
+    def _step(self, tensordict: TensorDictBase) -> TensorDictBase:
+        action = tensordict.get(self.action_key)
```
Comment (vmoens) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py:116]: Are actions always only a single tensor? 
This could break if you have multiple action tensors

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,601 @@
...
+                )
+        else:
+            tensordict_out = TensorDict._new_unsafe(
+                obs_dict,
+                batch_size=tensordict.batch_size,
+            )
+        if self.device is not None:
+            tensordict_out = tensordict_out.to(self.device)
+
+        self._curr_ep_len += 1
```
Comment (vmoens) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py:166]: I think that up until here everything is more or less like the `GymWrapper._step`, maybe `super()` would be more apporpriate?

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,601 @@
...
+        self._curr_reward_sum += reward
```
Comment (vmoens) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py:167]: if this is neede there's a transform that does that (RewardSum)

Comment (vmoens) [reply source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py:167]: We can append it by default using metaclasses

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py [hunk 10]
```diff
@@ -0,0 +1,601 @@
...
+        self._ep_len_buf = torch.where(~self._env_reset_mask, self._curr_ep_len.clone(), self._ep_len_buf)
+        self._ep_reward_buf = torch.where(~self._env_reset_mask, self._curr_reward_sum.clone(), self._ep_reward_buf)
+        if len(new_ids) > 0:
+            self._env_reset_mask[new_ids] = True
+            self._ep_len_buf[new_ids] = self._curr_ep_len[new_ids].clone()
+            self._ep_reward_buf[new_ids] = self._curr_reward_sum[new_ids].clone()
+
+        self._curr_ep_len[new_ids] = 0
+        self._curr_reward_sum[new_ids] = 0
+        tensordict.set("episode_length", self._ep_len_buf)
```
Comment (vmoens) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py:180]: here too we could use `StepCounter`

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py [hunk 11]
```diff
@@ -0,0 +1,601 @@
...
+from torchrl.objectives import ClipPPOLoss
+from torchrl.objectives.utils import _reduce
+from torchrl.record.loggers.wandb import WandbLogger
+from torchrl.trainers import Trainer
+from torchrl.trainers.trainers import LOGGER_METHODS
+
+from omni.isaac.lab.envs import ManagerBasedRLEnv
+
+
+class TorchRLEnvWrapper(GymWrapper):
```
Comment (vmoens) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py:35]: I think we can get around most of the extra features here with torchrl transforms and by telling GymWrapper that the env is auto-resetting - happy to help getting this done. 
That will be more robust long term, as this implementation relies on a bunch of private features of torchrl that we could modify in the future.

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py [hunk 12]
```diff
@@ -0,0 +1,601 @@
...
+
+    def iterator(self) -> Iterator[TensorDictBase]:
+        """Iterates through the DataCollector.
+
+        Yields: TensorDictBase objects containing (chunks of) trajectories
+
+        """
+        # The portion of the code handling cuda streams has been removed in this inherited method, which
+        # caused CUDA memory allocation issues with IsaacSim during env stepping.
+        total_frames = self.total_frames
```
Comment (vmoens) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py:304]: gotcha I think we can solve this by providing a `no_stream` arg in the collector - happy to make it a feature if you think that would help

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py [hunk 13]
```diff
@@ -0,0 +1,601 @@
...
+                # >>>          data0 = data
+                # >>>      elif i == 1:
+                # >>>          data1 = data
+                # >>>      else:
+                # >>>          break
+                # >>> assert data0["done"] is not data1["done"]
+                yield tensordict_out.clone()
+
+
+class ClipPPOLossWrapper(ClipPPOLoss):
```
Comment (vmoens) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py:356]: Can you comment on what has been changed here?
There are many versions of PPO and we probably don't cover exactly what you want it to do but that can be fixed!

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py [hunk 14]
```diff
@@ -0,0 +1,601 @@
...
+        td_out = td_out.named_apply(
+            lambda name, value: (
+                _reduce(value, reduction=self.reduction).squeeze(-1) if name.startswith("loss_") else value
+            ),
+            batch_size=[],
+        )
+        return td_out
+
+
+class TrainerWrapper(Trainer):
```
Comment (vmoens) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/wrappers/torchrl/torchrl_env_wrapper.py:470]: The trainer class hasn't been touched in a long time, as discussed I'm considering subclassing it for the various losses we have starting with PPO

**Review by vmoens — COMMENTED** [review]
Very quick review, happy to discuss each of these points

**Comment by AntoineRichard** [comment]
Hi @fyu-bdai — thank you so much for opening this PR, and for taking the time to contribute to Isaac Lab! 🙏

We're currently doing a big spring-clean of our pull request backlog, which had grown past 400 open PRs. Being honest: at that volume we stopped being able to give every contribution the attention it deserved, and this one is a case where we dropped the ball on getting back to you. That's on us, not on you.

**Why this PR is being closed:** Here is exactly what we found on this PR when we reviewed the backlog:

| | |
|---|---|
| **Opened** | 2024-10-08 (over a year ago) |
| **Last commit on the branch** | 2025-01-09 |
| **Last activity from the author** | over a year ago |
| **Target branch** | `main` — note that active development happens on `develop`; `main` only receives releases |
| **Review status** | Reviewed by vmoens, but it was never brought to a conclusion |
| **Merge status** | Conflicts with the base branch — it no longer applies cleanly |
| **Size** | 10 commit(s), 6 file(s) changed, +1200 / -0 |

It was picked up by the sweep because it targets `main` instead of `develop`, and it has been open for over a year. It was then put in the "close" bucket because the author has been silent for over a year — which is the signal we used to tell apart pull requests that are still being worked on from ones that have genuinely been set aside.

We deliberately did **not** close pull requests that were approved and ready to land, or that were small and clearly still fixing a live bug — there were 27 of those, and we are merging them rather than closing them.

To be completely clear — **this is not a judgement on the quality or the value of your work.** It's purely about getting the queue down to a size where we can actually review things properly and give contributors real, timely feedback.

**One thing worth knowing first:** Isaac Lab 3.0 has since shipped. `develop` is now at
3.0.0 while `main` is still at 2.3.2, and 3.0 restructured a fair amount — quaternion
conventions, actuator collections, data access, and the Isaac Sim extension imports all
changed. The [migration guide](https://github.com/isaac-sim/IsaacLab/blob/develop/docs/source/migration/migrating_to_isaaclab_3-0.rst)
is the place to start if you rebase.

It is also the honest reason a lot of these older pull requests stopped applying: the code
they touched moved or was rewritten underneath them.

**If this change is still relevant, please reopen this PR or open a fresh one against `develop`.** With the backlog down to a reviewable size, it should get a response a lot sooner than this one did.

Thanks again for contributing to Isaac Lab, and sorry for the long silence! 💚

---
🤖 *This comment was drafted with AI assistance as part of a maintainer-led sweep of the Isaac Lab pull request backlog. A maintainer is behind this cleanup — but if this closure looks wrong, it may well be, so please push back and we'll take another look.*
